### PR TITLE
Cn autograd debugging

### PR DESCRIPTION
**Pull request by cuong450**
Here is Python notebook showing inefficient gradient descent with the use of PopulationData_ag.py. I think you would be most interested in reviewing the relevant functions within PopulationData_ag.py:

- get_forecasted_data
- get_loss_given_truthdict_and_params
- get_grad_of_loss

### CODE REVIEW

File: cn_autograd_debugging/PopulationData_ag.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+from bs4 import BeautifulSoup
+import urllib
+
+
+
+
+
+
+def log_gamma_pdf(x,alpha,beta):
+    return np.log((beta**alpha)*(x**(alpha-1))*np.exp(-beta*x)/gamma_fcn(alpha))
```
Comment (michaelchughes) [inline cn_autograd_debugging/PopulationData_ag.py]: should do this all in log space directly, rather than compute pdf and then take the log

File: cn_autograd_debugging/PopulationData_ag.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    def get_loss_given_truthdict_and_params(self, params, truth_dict, lambda_reg):
+        # pop_states = ['infections', 'symptomatic','severe','hosp']
+        pop_states = ['hosp']
+        pred_dict = {'infections':[], 'symptomatic':[], 'severe':[], 'hosp':[],}
+        # print('get_loss_given_truthdict_and_params', truth_dict,'truth_dict')
+        for forecast_type in pop_states:
+            for t in range(len(truth_dict[forecast_type])):
+                self.params_for_training['t'] = t
+                self.params_for_training['forecast_type'] = forecast_type
+                pred_dict[forecast_type] += [self.get_forecasted_data(params)]
```
Comment (michaelchughes) [inline cn_autograd_debugging/PopulationData_ag.py]: incrementally growing the list will be crazy slow

you want to *preallocate* (we know for each channel we want to look at a fixed duration period of T days, so just make an array of size T)

File: cn_autograd_debugging/PopulationData_ag.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+        loss = 0
+        for i,k in enumerate(pop_states):
+            loss+=np.sum(np.abs(np.array(pred_dict[k]) - np.array(truth_dict[k]))*np.linspace(0.1, 1, num=len(pred_dict[k])) )*(i+1)
```
Comment (michaelchughes) [inline cn_autograd_debugging/PopulationData_ag.py]: several ways to make this much faster

(1) make the predictions and truth arrays to begin with, then you don't need to cast them (which might create copies and would be super expensive)
(2) create the array of weights from 0.1 to 1 ONCE, not each inner loop

Suppose you had:

pred_TK : 2D array of predictions, shape (T,K) == (n_timesteps, n_states)
truth_TK : 2D array of true values

You could do this whole computation in vectorized form:

```
weight_T1 = np.linspace(0.1, 1, T)[:,np.newaxis]
np.sum(weight_T1 * np.abs(pred_TK - truth_TK))
```

File: cn_autograd_debugging/PopulationData_ag.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        flow_sus_to_inf_list=[self.filtered_data['infections'][-1]]
+        Rt_list=[self.filtered_data['Rt'][-1]]
+        T_serial_list = [self.params['T_serial']]
+
+        #modify Rt based on government restrictionss
+        Rt_0 = self.filtered_data['Rt'][-1]
+        Rt_sf = self.future_data.filter_by([d.replace('-','') for d in dates_to_forecast], 'date').sort(['date'])['Rt'].to_numpy()
+
+        
+        for i,d in enumerate(dates_to_forecast):
```
Comment (michaelchughes) [inline cn_autograd_debugging/PopulationData_ag.py:200]: I'm pretty sure you can do this computation in vectorized way without a for loop

I'll try to mock something up and send a draft soon

Comment (michaelchughes) [reply cn_autograd_debugging/PopulationData_ag.py:200]: at very least, again do NOT grow a list one entry at a time, horribly inefficient

You want to preallocate and just fill the entries as you go

File: cn_autograd_debugging/PopulationData_ag.py (hunk 1 above)
Comment (cuong450) [reply cn_autograd_debugging/PopulationData_ag.py]: You mean using the log rules of multiplication and division? such that the resulting return value is a series of logs?

[alpha*np.log(beta)] + [(alpha-1)*np.log(x)] - [beta*x]   - np.log(gamma_fcn(alpha))

Is the reason just for numerical stability with extreme values?

**Comment by cuong450** [comment]
Hi Mike, can you take a look at autograd_for_poisson.py and tell me what I am doing wrong in trying to get the logpmf gradient?

File: cn_autograd_debugging/autograd_for_poisson.py [hunk 5]
```diff
@@ -1,7 +1,9 @@
...
+def logpmf_at_x(params,x):
+    return logpmf(x-params['loc'],params['mu'])
 params2= {'mu':2.0,'loc':5.0}
-print(logpmf(5.0-params2['loc'],params2['mu']))
-g2 = autograd.grad(logpmf)
-g2(5.0-params2['loc'],params2['mu'])
\ No newline at end of file
+print(logpmf_at_x(params2,5.0))
+g2 = autograd.grad(logpmf_at_x)
+g2(params2,5.0)
```
Comment (michaelchughes) [inline cn_autograd_debugging/autograd_for_poisson.py:9]: I'd do something like this, assuming you want the gradient of the logpmf wrt its "location" parameter.

# Import the poisson functionality
from autograd.scipy.stats import poisson as ag_poisson

# Remember the syntax: first arg is the integer value of the rv, second arg is the location value (scalar real)

```
In [37]: ag_poisson.pmf(0, 3.0)
Out[37]: 0.049787068367863944

In [38]: ag_poisson.pmf([0, 1, 2, 3, 4, 5, 6], 3.0).sum() #first 6 possible values take >90% of all probability mass
Out[38]: 0.9664914646911589
```

# Now get the gradient, make sure ag knows to take grad wrt the second argument (at index 1 not index 0)
 g = autograd.grad(poisson.logpmf, argnum=1)

# Evaluate the gradient

```
In [39]: g(6, 3.0)
Out[39]: 1.0

In [40]: g(1, 3.0)
Out[40]: -0.6666666666666667
```

File: cn_autograd_debugging/PopulationData_ag.py (hunk 4 above)
Comment (cuong450) [reply cn_autograd_debugging/PopulationData_ag.py:200]: Thanks for looking. The speed issue is now resolved, I removed a lot of the for loops of iterating day by day. 

Will push new version soon

File: cn_autograd_debugging/autograd_for_poisson.py (hunk 5 above)
Comment (cuong450) [reply cn_autograd_debugging/autograd_for_poisson.py:9]: g(1, 3.0)
I think this is taking the gradient with respect to mu. 

The challenge I am facing is taking the gradient with respect to  'location' and 'mu' at the same time. Which I would usually use a dictionary, but this hasn't been successful 

params_dict={mu:3.0,loc:5.0}
g(1, params_dict)


from this page:
https://github.com/HIPS/autograd/blob/master/autograd/scipy/stats/poisson.py
is this implying that autogrid is only able to take derivative with respect to mu and not with respect to location? where 'location' is suppose to be our population_model_predicted_number_of_hosp_admission

**Review by cuong450 — COMMENTED** [review]
for the  OVERFLOW.ipynb notebook which does a small run of gradient descent, you can see that on the first iteration, hospital predictions are actual values and not NaNs. but then I get the pink error box that says overflow encountered in exp of autograd vjp packages.

Pink RuntimeWarning:
"/Users/cuongnguyen/Desktop/cn_autograd_debugging/env/lib/python3.8/site-packages/autograd/tracer.py:48: RuntimeWarning: overflow encountered in exp
  return f_raw(*args, **kwargs)
/Users/cuongnguyen/Desktop/cn_autograd_debugging/env/lib/python3.8/site-packages/scipy/stats/_discrete_distns.py:763: RuntimeWarning: invalid value encountered in subtract
  Pk = special.xlogy(k, mu) - gamln(k + 1) - mu
/Users/cuongnguyen/Desktop/cn_autograd_debugging/env/lib/python3.8/site-packages/autograd/numpy/numpy_vjps.py:75: RuntimeWarning: invalid value encountered in double_scalars
  defvjp(anp.exp,    lambda ans, x : lambda g: ans * g)
"

In regards to using logpmf, I used the pre built autograd.scipy.special poisson logpmf
